### ISSUE TITLE
Intermittent failures due to memleaks detected in threads_test in sanitized builds

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#14437]
The recent addition of concurrent provider loading in the threads_test reveals new intermittent failures in sanitized builds. There are memleaks being detected. 

The problem was most probably already there in the code just the new test case is revealing it.

Shane had this comment:

@mattcaswell or @paulidale see the Sanitizer failure here? (I had this in my recent PR also)
I think test_threads was recently updated - I assume this is finding a race condition now?

This particular line in fipsprov.c indicates that something pretty bad is going on..
```
/* TODO(3.0): Tests will hang if this is removed */
(void)RAND_get0_public(libctx);
```

**Comment by t8m** [comment]
The issue for the TODO in fipsprov.c is #14364

**Comment by paulidale** [comment]
Weirdly, it's not a thread sanitiser issue.

**Comment by DDvO** [comment]
Coverity recently reported several issues with `test/threadstest.c` - maybe they are related.

**Comment by arminfuerst** [comment]
I also had a closer look at this, what I found out:
I could reproduce the error on my local machine with the test `test_multi_load` it `config` was run with parameter `enable-asan`.

**Comment by arminfuerst** [comment]
I'm not sure this is really related, I am mainly hit by the failing error of `90-test_memleak`.
To investigate, I quickly set up an Ubuntu 20.04 environment and I can reproduce the behavior there. I am pretty sure this is related to changes in the environment. When I can `valgrind test/memleaktest` I get no error (All heap blocks were freed -- no leaks are possible) what is certainly incorrect since this is leaking reliably. Additionally, I noticed following note:
ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
I did not succeed getting this running, neither by setting LD_PRELOAD nor by modifying the Makefile manually. I case I can make some tests, feel free to contact me.

**Comment by t8m** [comment]
Failing 90-test_memleak is a different issue and we do not see it in the CI. Please open a new issue with details - configure options and log of the memleak test failure.

**Comment by t8m** [comment]
Ah, that failure is appearing on Ubuntu 20.04 runners but not on 18.04 and the openssl project CI is still on 18.04. I'll investigate.

**Comment by arminfuerst** [comment]
> Ah, that failure is appearing on Ubuntu 20.04 runners but not on 18.04 and the openssl project CI is still on 18.04. I'll investigate.

On a branch on my fork I duplicated the non-caching to use 18.04 and 20.04 specifically. Here is a link to one of the results: https://github.com/arminfuerst/openssl/actions/runs/644269314
There is already some debug output. Shall I still create a new issue with additional details? I didn't have the time yet, I guess I will only find enough time this evening.

**Comment by t8m** [comment]
Fixed by merging PR #14497

**Comment by arminfuerst** [comment]
I saw there is already an issue (#14337), I'll continue there. I just confused the numbers (14337 / 14437 - that's why I started in this issue in the beginning...)
